### PR TITLE
TOOLS-2268 Add --mode=delete option

### PR DESCRIPTION
**Pull request by tfogo**
When mode is set to delete, mongoimport will delete existing documents in the database with matching documents from the import file. No operation will be performed with non-matching documents. By default, the `_id` field will be used to match documents. Use `--upsertFields` to specify the fields against which to match existing documents.

A couple of notes: 
- The original ticket requested "remove" mode. I thought that "delete" was more in line with the MongoDB API and also sounds a bit more serious (which is a good thing for something that will delete data).
- The `--upsertFields` option is now mis-named. I think `--matchFields` might be a better name.

### CODE REVIEW

**Comment by tfogo** [comment]
When I ran `dep ensure`, it added a bunch of missing files to the go driver package. It looks like it didn't change the revision (`d42ea72`), just those parts were missing in the repo for some reason. Any reason that would have happened?

**Comment by xdg** [comment]
I agree on the `--upsertField` comment.  Why don't you open a ticket about it and add it to the CLI rationalization Epic.  I'll look at the code tomorrow.

File: mongoimport/mongoimport.go [hunk 1]
```diff
@@ -490,7 +492,7 @@ readLoop:
 
 func (imp *MongoImport) updateCounts(result *mongo.BulkWriteResult, err error) {
 	if result != nil {
-		atomic.AddUint64(&imp.insertionCount, uint64(result.InsertedCount)+uint64(result.ModifiedCount)+uint64(result.UpsertedCount))
+		atomic.AddUint64(&imp.insertionCount, uint64(result.InsertedCount)+uint64(result.ModifiedCount)+uint64(result.UpsertedCount)+uint64(result.DeletedCount))
```
Comment (xdg) [inline mongoimport/mongoimport.go]: I think we should rename `insertionCount`.  What about "successCount" to mirror the existing "failureCount"?

File: mongoimport/mongoimport_test.go [hunk 2]
```diff
@@ -754,6 +765,43 @@ func TestImportDocuments(t *testing.T) {
 			}
 			So(checkOnlyHasDocuments(*imp.SessionProvider, expectedDocuments), ShouldBeNil)
 		})
+		Convey("CSV import with --mode=delete should succeed", func() {
+			// Fist import 3 documetns
```
Comment (xdg) [inline mongoimport/mongoimport_test.go]: Typos.

File: mongoimport/options.go [hunk 3]
```diff
@@ -78,7 +78,7 @@ type IngestOptions struct {
 	// "insert": Insert only, skip existing documents.
 	// "upsert": Insert new documents or replace existing ones.
 	// "merge": Insert new documents or modify existing ones; Preserve values in the database that are not overwritten.
-	Mode string `long:"mode" choice:"insert" choice:"upsert" choice:"merge" description:"insert: insert only. upsert: insert or replace existing documents. merge: insert or modify existing documents. (default: insert)"`
+	Mode string `long:"mode" choice:"insert" choice:"upsert" choice:"merge" choice:"delete" description:"insert: insert only. upsert: insert or replace existing documents. merge: insert or modify existing documents. (default: insert)"`
```
Comment (xdg) [inline mongoimport/options.go]: Need to add "delete" into description as well.

File: mongoimport/mongoimport_test.go [hunk 4]
```diff
@@ -754,6 +765,43 @@ func TestImportDocuments(t *testing.T) {
...
+			So(err, ShouldBeNil)
+
+			imp.InputOptions.Type = CSV
+			imp.InputOptions.File = "testdata/test_delete.csv"
+			fields = "_id,c,b"
+			imp.InputOptions.Fields = &fields
+			imp.IngestOptions.Mode = modeDelete
+			imp.IngestOptions.StopOnError = true
+			// Must specify upsert fields since option parsing is skipped in tests
+			imp.upsertFields = []string{"_id"}
```
Comment (xdg) [inline mongoimport/mongoimport_test.go]: Should add a test that uses a non-ID field to delete -- which gives me an interesting question in that case... are we doing "delete one" or are we doing "delete many"?  To be consistent with other modes, I think we should be intending this to be single-document -- and so we need to document that the update fields should uniquely identify a document.

Comment (tfogo) [reply mongoimport/mongoimport_test.go]: Yes it is "delete one".

File: mongoimport/mongoimport.go [hunk 5]
```diff
@@ -511,6 +513,8 @@ func (imp *MongoImport) importDocument(inserter *db.BufferedBulkInserter, docume
 			result, err = inserter.Insert(document)
```
Comment (tfogo) [inline mongoimport/mongoimport.go]: There's a neat little bug where if `--ignoreBlanks` is set, then mongoimport could fallback on insert mode. This causes `--mode delete` to insert documents.

E.g. 

```
./bin/mongoimport -c bar --file ./mongoimport/testdata/test_blanks.csv --type csv --fields a,b,c --ignoreBlanks --mode delete --upsertFields b -v
```

The issue is the `if selector == nil` line. I think the expected behaviour would be to just skip documents where we can't construct the filter.

**Comment by tfogo** [comment]
Will revendor before the merge to fix conflicts

File: mongoimport/mongoimport.go [hunk 6]
```diff
@@ -369,7 +371,7 @@ func (imp *MongoImport) ImportDocuments() (uint64, uint64, error) {
 // importDocuments is a helper to ImportDocuments and does all the ingestion
 // work by taking data from the inputReader source and writing it to the
 // appropriate namespace
-func (imp *MongoImport) importDocuments(inputReader InputReader) (numImported uint64, numFailed uint64, retErr error) {
+func (imp *MongoImport) importDocuments(inputReader InputReader) (uint64, uint64, error) {
```
[the remark this answers is not in the document]
Comment (tfogo) [reply mongoimport/mongoimport.go]: Good point, done!

File: mongoimport/mongoimport.go [hunk 7]
```diff
@@ -415,9 +419,9 @@ func (imp *MongoImport) importDocuments(inputReader InputReader) (numImported ui
 	}()
 
 	e1 := channelQuorumError(processingErrChan, 2)
```
Comment (tfogo) [inline mongoimport/mongoimport.go]: @xdg as an aside, why do we wait for 2 errors here and not 1?

File: Gopkg.toml [hunk 8]
```diff
@@ -29,11 +29,11 @@ required = ["github.com/3rf/mongo-lint"]
 
 [[constraint]]
   name = "github.com/mongodb/mongo-tools-common"
-  branch = "TOOLS-2268"
+  version = "v1.0.1"
```
Comment (xdg) [inline Gopkg.toml]: This should be "^v1.0.1" but master already has that so just don't lose it in the conflict resolution.  (No need to fix now as part of this PR.)

File: mongoimport/mongoimport_test.go [hunk 9]
```diff
@@ -792,16 +802,94 @@ func TestImportDocuments(t *testing.T) {
...
+			So(numProcessed, ShouldEqual, 2)
 			So(numFailed, ShouldEqual, 0)
 
 			expectedDocuments := []bson.M{
 				{"_id": int32(3), "c": 5.4, "b": "string"},
 			}
 			So(checkOnlyHasDocuments(*imp.SessionProvider, expectedDocuments), ShouldBeNil)
 		})
+		Convey("CSV import with --mode=delete and --upsertFields should succeed", func() {
+			// Fist import 3 documents
```
Comment (xdg) [inline mongoimport/mongoimport_test.go]: typo

**Pull request merged**